### ISSUE TITLE
Decide whether and how to make test environment configurable

### ISSUE THREAD

**Issue by tomjaguarpaw** [tomjaguarpaw/haskell-opaleye#288]
Dear @k0001, @xldenis, @phadej,

We currently have three open PRs aimed at making the test environment configurable

* https://github.com/tomjaguarpaw/haskell-opaleye/pull/210 @k0001 
* https://github.com/tomjaguarpaw/haskell-opaleye/pull/224 @xldenis 
* https://github.com/tomjaguarpaw/haskell-opaleye/pull/287 @phadej

We need to decide on a single approach and go for it.  If possible please provide links to other repositories demonstrating the state of the art in test configurability.  NB I am extremely disinclined to use environment variables since they are unpleasantly implicit.  However, given enough prior art I could be convinced otherwise.

**Comment by phadej** [comment]
There are three options:
- command line arguments, but it isn't really an option as test driver owns arguments
- config file
- env vars

IMHO the last one is most convenient (e.g. "how to run tests" with config file is more complicated). In fact, you can remove travis detection from the test suite, as you can set environment variables there *easily*.
You can write config file too, but it's less declarative in `.travis.yml`. For prioir art, I use environment variables in `github` tests, to pass in API credentials: https://github.com/phadej/github/blob/1ee5dc9cbfc1ae31c61ca0c92323fd55621e5ebf/spec/GitHub/UsersSpec.hs#L30

One more point, with environment variable it's much easier to configure multiple packages, especially convenient if they share the configuration bits (but you have to careful to normalise the envvar names). E.g. it's possible that Stackage would run tests at some point for packages requiring running services, there env-var approach would be a win.

**Comment by xldenis** [comment]
Maybe as a compromise we could use something like `dotenv`. My only concern against env variables is that I'm of the opinion that `stack test` should do the 'right thing' on it's own.

**Comment by tomjaguarpaw** [comment]
What problem are we trying to solve here exactly?  It seems like

1. We need to allow different users to configure their own test databases in a semi-permanent way
2. We need to be able to distinguish when it's Travis running the tests
3. Having indivivdual developers temporarily edit `connectInfo` in `Test.hs` is unsatisfactory because they have to ensure they don't add it to git by mistake.

Are there any other requirements or desires here?

**Comment by phadej** [comment]
@tomjaguarpaw I agree, and IMHO 1. and 2. are the same problem. Travis is "one user among others".

One compromise is both read config file and the env variable, the additional complexity is very small. (config file is more permanent way indeed, good point).

**Comment by xldenis** [comment]
> One compromise is both read config file and the env variable, the additional complexity is very small. 

That's exactly what `dotenv` provides.

**Comment by phadej** [comment]
I'm :+1: for http://hackage.haskell.org/package/dotenv, I can amend my PR

**Comment by tomjaguarpaw** [comment]
Hmm ... dotenv definitely sounds like a step backwards to me.  I want to get *away* from implicit mutable state, not towards it!

I'm sure I can be convinced eventually to *load* configuration from environment variables, but why pass it around my program within environment variables?

**Comment by phadej** [comment]
@tomjaguarpaw there's also http://hackage.haskell.org/package/dotenv-0.3.1.0/docs/Configuration-Dotenv-Parse.html, which doesnt't mutate environment

**Comment by saurabhnanda** [comment]
My 2 cents on how Rails does it. There's a `database.yml` which  looks like this:

```
production:
  host: <ip address comes here>
  port: <port comes here>
  username: <username comes here>
  password: <password comes here>

test:
  host: <ip address comes here>
  port: <port comes here>
  username: <username comes here>
  password: <password comes here>

developerA:
  host: <ip address comes here>
  port: <port comes here>
  username: <username comes here>
  password: <password comes here>

developerB:
  host: <ip address comes here>
  port: <port comes here>
  username: <username comes here>
  password: <password comes here>
```

And you have to start the app with `RAILS_ENV=developerA rails server` for the correct config to be picked up.

The downside of this approach is that every developer's local credentials will start creating noise in the global `database.yml` file. There are two workaround that I can think of:

* we don't have `developerA` and `developerB` environments but just on `development` environment and everyone is encouraged to use the same local creds.
* The app looks at `~/.opaleye/database.yml` (or suchlike) and merges it with `database.yml` during runtime and THEN uses the merged YML. This allows individual developers to have separate credentials without committing them to the central repo.

**Comment by phadej** [comment]
@saurabhnanda IMHO, too complicated for configuring single configuration string (IMHO dotenv solves ithat perfectly)

**Comment by xldenis** [comment]
> The downside of this approach is that every developer's local credentials will start creating noise in the global database.yml file.

The standard rails way is to provide a single `development` environment which avoids that noise, however, in this case I think `dotenv` addresses all of our concerns in a simpler manner.

**Comment by tomjaguarpaw** [comment]
@phadej That doesn't actually load the file though, does it?

**Comment by tomjaguarpaw** [comment]
I think this was fixed by 80513a584c1c9bee18c38b7b91b54688b3f6aa7d.
